promql/functions.go

```diff
@@ -79,22 +75,29 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
 		}
 		resultValue := lastValue - samples.Values[0].Value + counterCorrection
 
-		targetInterval := args[0].(*MatrixSelector).Range
-		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp)
-		if sampledInterval == 0 {
-			// Only found one sample. Cannot compute a rate from this.
-			continue
+		// Duration between first/last samples and boundary of range.
+		durationToStart := samples.Values[0].Timestamp.Sub(rangeStart).Seconds()
+		durationToEnd := rangeEnd.Sub(samples.Values[len(samples.Values)-1].Timestamp).Seconds()
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values))
+
+		// If the first/last samples are close to the boundaries of the range,
+		// extrapolate the result. This is as we expect that another sample
+		// will exist given the spacing between samples we've seen thus far,
+		// with an allowance for noise.
+		extrapolationThreshold := averageDurationBetweenSamples * 1.1
+		extrpolateToInterval := sampledInterval
+		if durationToStart < extrapolationThreshold {
+			extrpolateToInterval += durationToStart
+		}
+		if durationToEnd < extrapolationThreshold {
+			extrpolateToInterval += durationToEnd
+		}
```

Comment:
> Our goals are clearly aligned, we just come to opposite conclusions.

Not quite. I'm trying to get something that okay for all reasonable use cases, while I feel you're aiming for one (large) subset of use cases.

The reason it took me this long to get to this PR was as I needed time to sit down and consider all the edge and corner cases in which this would cause artifacts. The best case where the user's data is sampled, linear and continuous is all fine, but that's not the only case. We've already seen some of them, and they're not going to go away with this change.
Users who see artifacts are going to either stop using Prometheus due to the oddness, or add bog us down providing support - possibly rejecting Prometheus as too complicated in the process if the reasoning is beyond them. Support is much easier if it doesn't include notions like "half the median" as the numeracy of the average person is to put it mildly not good (and even for those you'd expect to be above average, a lecturer in one very well regarded US university avoids fractions with freshmen).

If we know we're going to have artifacts and incorrect answers, we need to keep the math and algorithms which are accordingly exposed to users simple. This isn't just a tradeoff in the technical space, we need to make sure that slightly better numbers in some cases (which as you say are fairly incremental in benefit) doesn't end up with a net loss in users due to them being scared away by math by other cases.